Packages/com.chisel.editor/Chisel/Editor/Editor/MaterialBrowser/ChiselMaterialBrowserUtilities.cs

```diff
@@ -0,0 +1,242 @@
+            {
+                e.Dispose();
+                e = null;
+            } );
+
+            materials.Clear();
+
+            // exclude the label search tag if we arent searching for a specific label right now
+            string search = usingLabel ? $"l:{searchLabel} {searchText}" : $"{searchText}";
+
+            string[] guids = AssetDatabase.FindAssets( $"t:Material {search}" );
+
+            // assemble preview tiles
+            foreach( string id in guids )
+            {
+                ChiselMaterialBrowserTile browserTile = new ChiselMaterialBrowserTile( id );
+
+                if( labels != null )
+                {
+                    // add any used labels we arent currently storing
+                    foreach( string label in browserTile.labels )
+                    {
+                        if( !labels.Contains( label ) )
+                            labels.Add( label );
+                    }
+                }
+
+                // check each entry against a filter to exclude certain entries
+                if( IsValidEntry( browserTile ) )
+                {
+                    // if we have the material already, skip, else add it
+                    materials.Add( browserTile );
+                }
+            }
+
+            AssetPreview.SetPreviewTextureCacheSize( materials.Count + 1 );
+
+            models = Object.FindObjectsOfType<ChiselModel>().ToList();
+
+            PopulateUsedMaterials( ref usedMaterials, ref models, searchLabel, searchText );
+        }
+
+        private static void PopulateUsedMaterials( ref List<ChiselMaterialBrowserTile> tiles, ref List<ChiselModel> models, string searchLabel, string searchText )
+        {
+            tiles.ForEach( e =>
+            {
+                e.Dispose();
+                e = null;
+            } );
+
+            tiles.Clear();
+
+            foreach( ChiselModel m in models )
+            {
+                MeshRenderer[] renderMeshes = m.generated.meshRenderers;
+
+                if( renderMeshes.Length > 0 )
+                    foreach( MeshRenderer mesh in renderMeshes )
+                    {
+                        foreach( Material mat in mesh.sharedMaterials )
+                        {
+                            if( mat != null )
+                                if( mat.name.Contains( searchText ) || MaterialContainsLabel( searchLabel, mat ) )
+                                {
+                                    AssetDatabase.TryGetGUIDAndLocalFileIdentifier( mat, out string guid, out long id );
+
+                                    tiles.Add( new ChiselMaterialBrowserTile( guid ) );
+                                }
+                        }
+                    }
+            }
+        }
+
+        private static bool MaterialContainsLabel( string label, Material material )
+        {
+            string[] labels = AssetDatabase.GetLabels( material );
+
+            if( labels.Length > 0 )
```

Comment:
I'd prefer it if you put brackets around the foreach in this if, it can prevent some accidental issues

Reply:
Or just early out 
if( labels.Length == 0 ) return false;

Reply:
Fixed